### PR TITLE
Be more obvious when suggesting dereference

### LINKED ISSUE #45925: Unhelpful (incorrect?) match_default_bindings hint
Here is the whole error output:

```
error: non-reference pattern used to match a reference (see issue #42640)
   --> src/coord.rs:194:23
    |
194 |             .filter(|&(ref k, _)| self.hash(k) == server)
    |                       ^^^^^^^^^^ help: consider using: `&(ref k, _)`
    |
    = help: add #![feature(match_default_bindings)] to the crate attributes to enable

error: aborting due to previous error

error: Could not compile `kv_2pc`.

To learn more, run the command again with --verbose.
```

Specifically, notice the hint: 
```
help: consider using: `&(ref k, _)`
```
which suggests doing something I am already doing 😛 ...

The code snippet in question:

```rust
    fn send_prepare(&self, server: &str, txn: TxnId, staged: &Vec<(String, String)>) -> Flags {
        // Only send to the server the items on that server!
        let mut staged = staged
            .iter()
            .filter(|&(ref k, _)| self.hash(k) == server)
            .collect();

        retry_forever!{
            do {
                //     SendToServer(s, {Prepare})
                let server = SyncClient::connect(server, client::Options::default()).unwrap();
                server.send_to_server(txn, Flags::PREPARE, Some(staged.clone()))
            },

            retry if Err(tarpc::Error::Io(_)),

            match {
                Ok(flags) => flags,
                err => panic!("Error with (de)serialization: {:?}", err)
            }
        }
    }
```

Changing `.filter(|&(ref k, _)| self.hash(k) == server)` to `.filter(|pair| self.hash(&pair.0) == server)` sidesteps the problem.

```
$ rustc --version --verbose
rustc 1.23.0-nightly (a35a3abcd 2017-11-10)
binary: rustc
commit-hash: a35a3abcda67a729edbb7d649dbc663c6feabd4c
commit-date: 2017-11-10
host: x86_64-unknown-linux-gnu
release: 1.23.0-nightly
LLVM version: 4.0
```

### PR DESCRIPTION
**Pull request by estebank**
Include `&` span when suggesting dereference on a span that is already a reference:

```
error: non-reference pattern used to match a reference (see issue #42640)
  --> dont-suggest-dereference-on-arg.rs:16:19
   |
16 |         .filter(|&(ref a, _)| foo(a))
   |                  ^^^^^^^^^^^ help: consider using: `&&(ref k, _)`
   |
   = help: add #![feature(match_default_bindings)] to the crate attributes to enable
```

Fix #45925.

### CODE REVIEW

**Comment by mark-i-m** [comment]
I'm still not quite sure I understand the error... Isn't `&(ref k, _)` a reference pattern used to match a reference? What does it mean by "non-reference pattern"?

**Comment by estebank** [comment]
*This PR is not a full solution, do not merge.*

@mark-i-m the type checker sees the closure's argument as a pattern match, I believed that behavior was introduced as part of [RFC 2005](https://github.com/rust-lang/rfcs/blob/master/text/2005-match-ergonomics.md) (tracked in #42640).

**Comment by mark-i-m** [comment]
It is a pattern match, though, right? I was under the impression that in Rust all arguments to functions and closures are actually patterns, which can bind new names, just as in an `if let` or `match`. Is this not true?

Another way of asking the question is to ask if `&(ref k, _)` is actually wrong. I believe the types are correct, and I don't think I was violating the borrow rules or anything. So IIUC I should not have gotten any error at all; it should have just compiled, right?

**Comment by mark-i-m** [comment]
Ah, I see. The correct pattern would be `&&(ref k, _)` but the "type mismatch" error is covered over by the "match bindings" error... So the code should not have compiled...

**Comment by mark-i-m** [comment]
Hmm... so the hint was actually on the right track. It suggests replacing `(ref k, _)` with `&(ref k, _)` in `&(ref k, _)`, which would have lead to `&&(ref k, _)`. It is just not at all clear that's what it means...

**Comment by arielb1** [comment]
@estebank 

Can't we just adjust our error message until it includes all the surrounding reference patterns, so we'll get this:
```Rust
error: non-reference pattern used to match a reference (see issue #42640)
   --> src/coord.rs:194:23
    |
194 |             .filter(|&(ref k, _)| self.hash(k) == server)
    |                      ^^^^^^^^^^^ help: consider using: `&&(ref k, _)`
    |
    = help: add #![feature(match_default_bindings)] to the crate attributes to enable

error: aborting due to previous error

error: Could not compile `kv_2pc`.

To learn more, run the command again with --verbose.

```

**Comment by estebank** [comment]
@arielb1 yeah, I think that's a better approach.

**Comment by kennytm** [comment]
Hi @estebank, what is the status of this PR?

**Comment by estebank** [comment]
@kennytm @arielb1 I've modified the PR to do some span fiddling to see if there's a `&` right before. I'm not entirely happy with doing that, but it should be enough to handle the presented situation.

**Comment by mark-i-m** [comment]
@estebank 
>  I'm not entirely happy with doing that

What problems do you foresee with it?

File: src/libsyntax_pos/lib.rs [hunk 1]
```diff
@@ -159,6 +159,18 @@ impl Span {
...
     }
 
+    /// Returns a new span representing the previous character after the start-point of this span
+    pub fn prev_point(self) -> Span {
+        let span = self.data();
+        let span_lo = span.lo.0;
+        let lo = if span_lo == 0 {
+            0
+        } else {
+            span_lo - 1
```
Comment (mark-i-m) [inline src/libsyntax_pos/lib.rs]: Doesn't this mean we now have the same problem with the following case:
```rust
let x = &&&0;
let &&x = x;
```

IIUC, the error message would suggest using `&&`, right?
```

File: src/libsyntax_pos/lib.rs [hunk 2]
```diff
@@ -159,6 +159,18 @@ impl Span {
         Span::new(BytePos(lo), BytePos(lo), span.ctxt)
     }
 
+    /// Returns a new span representing the previous character after the start-point of this span
+    pub fn prev_point(self) -> Span {
+        let span = self.data();
+        let span_lo = span.lo.0;
+        let lo = if span_lo == 0 {
+            0
```
Comment (arielb1) [inline src/libsyntax_pos/lib.rs]: This looks likely to introduce an infinite loop in *some* case, where we go up to the start of the file. You can return an `Option<Span>` instead.

**Comment by arielb1** [comment]
@estebank 

Please don't fiddle with spans, it causes very weird situations with macros. You can use `tcx.hir.get_parent_node` to go up the pattern tree and find the reference patterns.

File: src/librustc_typeck/check/_match.rs [hunk 3]
```diff
@@ -120,15 +120,33 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                         .pat_adjustments_mut()
                         .insert(pat.hir_id, pat_adjustments);
                 } else {
+                    let mut ref_sp = pat.span;
+                    let mut id = pat.id;
+                    loop {  // make span include all enclosing `&` to avoid confusing diag output
```
Comment (arielb1) [inline src/librustc_typeck/check/_match.rs]: nit: s/loop/while let/

Comment (arielb1) [reply src/librustc_typeck/check/_match.rs]: ok a while-let would be ugly too. Why can't we have a do-while loop?

**Comment by kennytm** [comment]
@bors r-

CI failed, the UI test `ui/suggestions/dont-suggest-dereference-on-arg.rs`’s output needs to be updated.

File: src/test/ui/suggestions/dont-suggest-dereference-on-arg.rs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+// except according to those terms.
+
+fn foo(s: &str) -> bool { true }
+
+fn main() {
+    let x = vec![(String::new(), String::new())];
+    x.iter()
+        .filter(|&(ref a, _)| foo(a))
+        //~^ ERROR non-reference pattern used to match a reference
+        //~| HELP consider using a reference
```
Comment (kennytm) [inline src/test/ui/suggestions/dont-suggest-dereference-on-arg.rs]:
```
[00:47:09] error: /checkout/src/test/ui/suggestions/dont-suggest-dereference-on-arg.rs:16: unexpected help message: '16:18: 16:29: add #![feature(match_default_bindings)] to the crate attributes to enable'
```

Needs one more `//~| HELP`.

**Pull request merged**